rir/src/compiler/opt/force_dominance.cpp

```diff
@@ -136,15 +133,63 @@ struct ForcedBy {
             }
         }
 
-        if (eagerFunction != NotEagerFunction &&
-            eagerFunction != other.eagerFunction) {
-            eagerFunction = NotEagerFunction;
+        if (forceArgsBeforeEffect && !other.forceArgsBeforeEffect) {
+            forceArgsBeforeEffect = false;
+            res.update();
+        }
+
+        if (!ambiguousForceOrder && other.ambiguousForceOrder) {
+            ambiguousForceOrder = true;
             res.update();
         }
 
+        if (argumentForceOrder != other.argumentForceOrder) {
+            auto mySize = argumentForceOrder.size();
+            auto otherSize = other.argumentForceOrder.size();
+            auto common = mySize;
+
+            if (mySize > otherSize) {
+                argumentForceOrder.resize(otherSize);
+                ambiguousForceOrder = true;
+                common = otherSize;
+                res.update();
+            } else if (!ambiguousForceOrder && otherSize > mySize) {
+                ambiguousForceOrder = true;
+                res.update();
+            }
+
+            for (size_t i = 0; i < common; ++i) {
+                if (argumentForceOrder[i] != other.argumentForceOrder[i]) {
+                    argumentForceOrder.resize(i);
+                    ambiguousForceOrder = true;
+                    res.update();
+                    break;
+                }
+            }
+        }
+
         return res;
     }
 
+    bool maybeForced(size_t i) const {
+        for (auto f : argumentForceOrder) {
+            if (f == i)
```

Comment:
This is to check if the given argument id `i` has already been forced?

Might be a bit heavy to bring in `<algorithm>` and `std::find`, but I think that would be slightly more clear. I definitely went "huh?" when I saw `if (f == i)`. Either way, a comment would help.